### ISSUE TITLE
Using Bond Push UDP Protocol

### ISSUE THREAD

**Issue by jvmahon** [dcmeglio/hubitat-bond#11]
It looks like the Bond V2 Protocol implements a push protocol (BPUP) that can be used instead of polling. Have you considered using BPUP instead of polling the Bond hub?  I may look into this as a way of learning groovy programming, but I assume you'd be able to figure it out much faster, so if you're already considering it yourself (or have hints on how to do this in your code), please let me know.

**Comment by dcmeglio** [comment]
Hubitat does not support unsolicited UDP messages so, unfortunately, this cannot be implemented until Hubitat adds that support.

**Comment by apwelsh** [comment]
Since UDP is sessionless, I wonder what they require to consider the UDP message solicited.  Is it possible to send a UDP packet out periodically, to open up to UDP port, and keep it open?

**Comment by dcmeglio** [comment]
No. Hubitat only receives a single UDP response and will ignore the rest. Unfortunately this is not possible with Hubitat. I’ve already confirmed this with the Hubitat staff.

**Comment by dcmeglio** [comment]
Hopefully they make it possible at some point. BPUP actually came up from a suggestion I made to Bond to do webhooks. Unfortunately they went down the UDP road making it unsupportable for the moment.

**Comment by chrismerck** [comment]
Comparing to mobile platforms: Android and iOS do not allow listening to an arbitrary UDP port, so the "broadcast" method doesn't work. However, you can open a UDP socket with Bond as the destination, and the OS will assign a random source port. There are plenty of ports to go around (~30000), even iOS allows this. All reply messages from the Bond go to that random source port and get forwarded to the application.

I would think that Hubitat could sustainably implement similar support for drivers. Webhooks doesn't scale well for deeply embedded products.

**Comment by dcmeglio** [comment]
@chrismerck it's definitely possible for Hubitat to add this support, however, when I spoke to the HE team, they basically said it's very low on their priority list so I'm not holding my breath. There are actually several integrations that would benefit from this so it's not just Bond. Not much I can do though, I made the suggestion. I'm sure it couldn't hurt if you as a fellow device manufacturer proposed it too!

The benefit of webhooks would have been that I wouldn't have to still, essentially, poll with keep alive messages. You'd only have to open the HTTP connection for those few times where there was actually a state change instead of us having to have communications every X seconds. That said, you know way more about how Bond works internally to know what's possible!
